### ISSUE TITLE
how to know value for tuner.setLoopInterval(loopInterval)

### ISSUE THREAD

**Issue by mcarosh** [jackw01/arduino-pid-autotuner#10]
Hi jackw01
as topic stated, how do I know the correct value for tuner.setLoopInterval(loopInterval)? I am using an Arduino Due and want to integrate PID-Autotuner but fail with this setting :/. Do I need to calculate this setting according to the looptime (processor speed Due)?
Furthermore I struggle to integrate PID-Autotuner by using a button. Which command do I need to call the calibration function
Is it this tuner.startTuningLoop(micros())? and do I have the possibility to indicate the state of the function? Something like
if (tunePid() == true)
{digital write (Pin13, 1)}
Maybe you can give some hints for solving these problems?!
So far thanks for sharing this library :)
regards
mcarosh

**Comment by mcarosh** [comment]
And sorry if there is a double post???
I already asked this question yesterday but today there is no issue open???

**Comment by BigMuscle85** [comment]
I have the similar question. What's the point in setLoopInterval when tunePID requires the actual time?

And does it have to be in microseconds or in the same time unit as our PID regulator? I tried using microseconds and also microseconds in our PID regulator and it computes really huge integral part that make large overshoot even with ZNModeNoOvershoot.

**Comment by mcarosh** [comment]
Hi bigmuscle85
What I did to solve this problem was to directly measure the loop-time.
In my case:

unsigned long timeBegin        = micros();
unsigned long timeEnd          = micros();
unsigned long duration ;

void loop(){

 timeBegin = micros();

 tuner.setLoopInterval(duration);
 .
 .
 .
  timeEnd = micros();
  duration = timeEnd - timeBegin;}

I´ve put this into the void loop because then i do not need to care about any change in my code
this time will always be actual (at the cost of processor-power but I use the Due:)
To me that worked
Hope for you too
Regards
mcarosh

**Comment by jackw01** [comment]
@mcarosh this is not correct. The tuning loop needs to run at a fixed rate (see the example code) and this rate needs to be the same as the rate the PID control loop will run at. Calling setLoopInterval during the loop will only work properly if the loop is being rate limited.

**Comment by BigMuscle85** [comment]
@mcarosh Thanks. I tried this way and it seems to work. I also noticed that the time units (s, ms, us) does not matter in this tuner (seconds require floats), but "dt" in PID control loop must be in seconds.

The PID control loop in our application does not run at fixed rate. Although it has sleep(500ms), it takes some additional non-fixed time to read the sensor and set up the new value. Total time is about 1200-1400 ms between subsequent PID calls.

ZNModeNoOvershoot still gives a large (about 10°C) overshoot. Is there some way to fine-tune the computed values except trial and error?

**Comment by mcarosh** [comment]
@jackw01 
thanks for your correction
I start tuner via button (if statement) and as i understand the logic then it is a fixed rate, isn´t it? And this was a solution to get the correct Loopinterval. To me it´s working but do you suggest to measure time once and put it into void setup???
regards
mcarosh

**Comment by mcarosh** [comment]
@BigMuscle85 
If you are able to display your values on a graph (maybe serial monitor) you will find tutorials (just google) how to manually tune PID.
And in my opinion this is the best method to get out best results. I also do the fine-tuning manually.
regards

**Comment by BigMuscle85** [comment]
@mcarosh I just used ZNModeNoOvershoot mode and reduced the computed Ki to its quarter. Now it is much better but still not perfect. Originally I just wanted to know whether this overshoot cannot be caused by improperly set loopInterval, but now I know that fixed loop interval or dynamically computed during the loop makes no difference.
